[quant] Member variable access in quantizable LSTM

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64834
* __->__ #64833

The quantizable LSTM stores it's weight/bias variables deeper in the hierarchy. To access the weights/biases, one must either use `module.named_parameters()` or dive deeper (s.a. `module.layers[0].layer_fw.cell.weight`). However, in the original `nn.LSTM`, those variables are accessible from the top level (i.e. `module.weight_ih_l0`).

This PR creates an alias on the top level to be able to access the weights/biases directly from the top. Note that because during the `prepare`/`convert` steps those variables change ID, we have to "reattach" them on the top level. Hence, the support method `_set_weight_bias_variables`, which is called during construction, `from_float`, and `from_observed`.

Major note: Once the module is quantized, the weight/bias variables are packed. That means that accessing them directly is impossible, as the variables return unpacking functions.

Differential Revision: [D30870715](https://our.internmc.facebook.com/intern/diff/D30870715/)

Commit:
[quant] Member variable access in quantizable LSTM

The quantizable LSTM stores it's weight/bias variables deeper in the hierarchy. To access the weights/biases, one must either use `module.named_parameters()` or dive deeper (s.a. `module.layers[0].layer_fw.cell.weight`). However, in the original `nn.LSTM`, those variables are accessible from the top level (i.e. `module.weight_ih_l0`).

This PR creates an alias on the top level to be able to access the weights/biases directly from the top. Note that because during the `prepare`/`convert` steps those variables change ID, we have to "reattach" them on the top level. Hence, the support method `_set_weight_bias_variables`, which is called during construction, `from_float`, and `from_observed`.

Major note: Once the module is quantized, the weight/bias variables are packed. That means that accessing them directly is impossible, as the variables return unpacking functions.

Differential Revision: [D30870715](https://our.internmc.facebook.com/intern/diff/D30870715/)

[ghstack-poisoned]

z-a-f commented:
@albanD THis is a follow up on the comment you had in the older PR.

Commit:
Update on "[quant] Member variable access in quantizable LSTM"


The quantizable LSTM stores it's weight/bias variables deeper in the hierarchy. To access the weights/biases, one must either use `module.named_parameters()` or dive deeper (s.a. `module.layers[0].layer_fw.cell.weight`). However, in the original `nn.LSTM`, those variables are accessible from the top level (i.e. `module.weight_ih_l0`).

This PR creates an alias on the top level to be able to access the weights/biases directly from the top. Note that because during the `prepare`/`convert` steps those variables change ID, we have to "reattach" them on the top level. Hence, the support method `_set_weight_bias_variables`, which is called during construction, `from_float`, and `from_observed`.

Major note: Once the module is quantized, the weight/bias variables are packed. That means that accessing them directly is impossible, as the variables return unpacking functions.

Differential Revision: [D30870715](https://our.internmc.facebook.com/intern/diff/D30870715/)

[ghstack-poisoned]

vkuzo reviewed (commented):
@z-a-f , could we update the PR summary with the exact user flow that this PR enables (a code sample)?  It's described in the comments but not obvious for someone who doesn't work with LSTMs.

z-a-f commented:
> @z-a-f , could we update the PR summary with the exact user flow that this PR enables (a code sample)? It's described in the comments but not obvious for someone who doesn't work with LSTMs.

I am not sure I understand -- this PR reintroduces deleted getters for the weight and bias. There is no user flow change.

Commit:
Update on "[quant] Member variable access in quantizable LSTM"


The quantizable LSTM stores it's weight/bias variables deeper in the hierarchy. To access the weights/biases, one must either use `module.named_parameters()` or dive deeper (s.a. `module.layers[0].layer_fw.cell.weight`). However, in the original `nn.LSTM`, those variables are accessible from the top level (i.e. `module.weight_ih_l0`).

This PR creates an alias on the top level to be able to access the weights/biases directly from the top. Note that because during the `prepare`/`convert` steps those variables change ID, we have to "reattach" them on the top level. Hence, the support method `_set_weight_bias_variables`, which is called during construction, `from_float`, and `from_observed`.

Major note: Once the module is quantized, the weight/bias variables are packed. That means that accessing them directly is impossible, as the variables return unpacking functions.

Differential Revision: [D30870715](https://our.internmc.facebook.com/intern/diff/D30870715/)

[ghstack-poisoned]

Commit:
Update on "[quant] Member variable access in quantizable LSTM"


The quantizable LSTM stores it's weight/bias variables deeper in the hierarchy. To access the weights/biases, one must either use `module.named_parameters()` or dive deeper (s.a. `module.layers[0].layer_fw.cell.weight`). However, in the original `nn.LSTM`, those variables are accessible from the top level (i.e. `module.weight_ih_l0`).

This PR creates an alias on the top level to be able to access the weights/biases directly from the top. Note that because during the `prepare`/`convert` steps those variables change ID, we have to "reattach" them on the top level. Hence, the support method `_set_weight_bias_variables`, which is called during construction, `from_float`, and `from_observed`.

Major note: Once the module is quantized, the weight/bias variables are packed. That means that accessing them directly is impossible, as the variables return unpacking functions.

Differential Revision: [D30870715](https://our.internmc.facebook.com/intern/diff/D30870715/)

[ghstack-poisoned]
